### ISSUE TITLE
Facebook App instead of webview 

### ISSUE THREAD

**Issue by atrimn** [facebook/react-native-fbsdk#471]
`react-native-fbsdk` is a wrapper around the Facebook SDKs for iOS and Android. GitHub Issues in the `facebook/react-native-fbsdk` repository are used exclusively for tracking bugs in the React Native wrapper. GitHub may not be the ideal place to ask questions about using React Native, but we have compiled a list of resources that should help.

### Report a bug in the Facebook SDK for iOS or Android

`react-native-fbsdk` is a wrapper around the Facebook SDKs for iOS and Android. If you've found a bug in the Facebook SDK itself, please visit https://developers.facebook.com/bugs/.

### Get help using `react-native-fbsdk` in your app

Many members of the community use Stack Overflow to ask questions. Read through the [existing questions](https://stackoverflow.com/questions/tagged/react-native-fbsdk?sort=frequent) tagged with **react-native-fbsdk** or [ask your own](http://stackoverflow.com/questions/ask?tags=react-native-fbsdk)!

### Get help with your React Native app

There are a lot of developers around the world using React Native. You can find a list of community resources at http://facebook.github.io/react-native/help

### Chat with React and React Native community members

If you need an answer right away, check out the [Reactiflux Discord](https://discord.gg/0ZcbPKXt5bZjGY5n) community. There are usually a number of React Native experts there who can help out or point you to somewhere you might want to look.
I cant seem to figure this one out. I have the facebook app installed on my ios device but when i use the login manager to log in it opens up a webview instead of the facebook app.

**Comment by SnaiNeR** [comment]
me too, only on ios
but on android works via app

**Comment by StevenMasini** [comment]
Yes it works for me on Android if I put the behavior in uppercase, but doesn't work on iOS no matter if the behavior is in uppercase or lowercase.

```js
import { LoginManager } from 'react-native-fbsdk';
import { Platform } from 'react-native';

let result;
try {
  let behavior = Platform.OS === 'ios' ? 'native' : 'NATIVE_ONLY';
  LoginManager.setLoginBehavior(behavior);
  result = await LoginManager.logInWithReadPermissions(['public_profile', 'user_birthday', 'user_gender', 'user_friends']);
} catch (err1) {
  try {
    let behavior = Platform.OS === 'ios' ? 'browser' : 'WEB_ONLY';
    LoginManager.setLoginBehavior(behavior);
    result = await LoginManager.logInWithReadPermissions(['public_profile', 'user_birthday', 'user_gender', 'user_friends']);
  } catch (err2) {
    // Facebook login failed with both native app and web browser
  }
}
if (result && !result.isCancelled) {
  //  Facebook login succeed, perform your login here
}
```

Nevertheless yesterday we received a warning from Facebook, which ask us to use the latest version of their SDK on iOS. But I double checked and we already do use the latest version `4.40.0`. 
I don't know if anyone else has received this email.

Apparently their warning might be related to the fact that the Native App doesn't launch properly in our Facebook login on iOS. See https://developers.facebook.com/policy#7-2-photo

**Comment by StevenMasini** [comment]
After putting some breakpoints in the source code of the native iOS SDK I succeed to track down the issue to the method `- (void)logInWithBehavior:(FBSDKLoginBehavior)loginBehavior` in `FBSDKLoginManager` where the `[FBSDKInternalUtility isFacebookAppInstalled]` return `false` even if the app is installed on the device and even if the user is connected to the app.

I opened a ticket on the bug report platform of Facebook and I am waiting for their reply.

https://developers.facebook.com/bugs/1541604222639302/

**Comment by atrimn** [comment]
@StevenMasini i think its not a bug facebook
thinks its better ux to have the webview

**Comment by Sixbitunder** [comment]
@atrimn I don't think so, in this way, the user always has to fill his login details in the browser. This is really bad for the user's experience especially if they have facebook app installed on their phones. In fact facebook block your login app because they don't want this.. And it's absurd! They ask for something that actually you cannot achieve in any way on iOS anymore. I've read that from iOS 11 the 'native' behavior started to point to webview and not to facebook app. Probably is something related to Apple and their new restrictions and this probably could be the reason why `[FBSDKInternalUtility isFacebookAppInstalled]` always returns `false`

**Comment by StevenMasini** [comment]
@atrimn Facebook don't think that the webview is a better approach. @Sixbitunder is right.

We received an email from Facebook warning us that our iOS version trigger the webview instead of the native app. They asked us to update the code to launch the native app, but there is this bug with `react-native-fbsdk`.

**Comment by RevanScript** [comment]
Any updates on this? It's not a minor bug that can be just ignored.

**Comment by StevenMasini** [comment]
@RevanScript I opened a ticket on their developer platform, but for now, still now news. You can follow the progression here

https://developers.facebook.com/bugs/1541604222639302/

**Comment by poptocrack** [comment]
Hi,
Quick update from the Facebook team:

> Hi John,
> We just got back from our team and this is unfortunately a won't fix for now. We'll keep tracking it internally and update whenever this gets fixed.
> 
> Please let me know if you have any other questions in the meantime.
> 
> Best,
> Gunay

**Comment by gsunsnackv** [comment]
So @StevenMasini , FB ask you to use their app instead of webview, but refuse to fix their bug that always says the app is not installed? What a bunch of assholes!

**Comment by StevenMasini** [comment]
@gsunsnackv Yeah that was the irony of the situation. But I have checked our log in with Facebook recently, and it seems to be fixed now. 

At least it is fixed on my app after updating to the latest version of `react-native-fbsdk`.

Maybe others could confirm ?

**Comment by HelloAfrica** [comment]
@StevenMasini so the current state of your ios app allows for facebook login to happen with the native app and not the browser? Are you sure you weren't logged in via the browser the last time you checked?

**Comment by StevenMasini** [comment]
@HelloAfrica I just checked now on my app, it does open a web page first, but then inside the web page it gives me the option to log in via the native app.

**Comment by HelloAfrica** [comment]
@StevenMasini could you share a screenshot?

**Comment by belal-mazlom** [comment]
That was default behaviour on iOS
Open web page then ask to log via native app or from webview
I think it should open native app directly without open web page

But checking source code I think there is not supported on iOS  😕
https://github.com/facebook/react-native-fbsdk/blob/master/js/FBLoginManager.js#L50

**Comment by bigorangemachine** [comment]
I totally ripped off @StevenMasini snippet (thanks!).  Tweaked it a little after trying a reduce async/await (come on es-next!).

I also noticed the `README.md` is showing the master not the `stable`.  If you are [pre react-native 0.60.0 I got hung up on this being missing](https://github.com/facebook/react-native-fbsdk/blob/831d88769cf94f11e5c07536a325953791d5b0f6/README.md)

With:
**react-native-fbsdk**: `0.10.1`
**react-native**: `0.59.9`

I'm leveraging an environment variable as the emulators have varying levels of support for the facebook app.  Which oddly enough my alpha didn't use `LoginManager.setLoginBehavior()` and it worked on a real-life iPad.  Once I applied 

In my `README.md`
```
* `FB_LOGIN_FORCE`: Can be `none`, `both`, `web`, `native` but defaults to `both`
```

In `src/App.js` -> populate `process.env` with values from `Config`
```
import Config from 'react-native-config';

Object.assign(process.env, Object.keys(Config).reduce((acc, key) => {
  if (typeof Config[key] === 'string' && !process.env[key]) {
    acc[key] = Config[key];
  }

  return acc;
}, {}));

export default process.env;
```

In my `src/utils/facebookSDK.js`
```
import FBSDK from 'react-native-fbsdk';

const { LoginManager } = FBSDK;
const fbLoginWebview = async (behavior = null) => {
  let auth;

  try {
    // falsey behavior should not set a mode
    if (behavior === 'native') {
      LoginManager.setLoginBehavior(Platform.OS === 'ios' ? 'native' : 'NATIVE_ONLY');
    } else if (behavior === 'web') {
      LoginManager.setLoginBehavior(Platform.OS === 'ios' ? 'browser' : 'WEB_ONLY');
    }

    auth = await LoginManager.logInWithPermissions(['public_profile']);
  } catch (caughtError) {
    // Facebook login failed -> Log for the local development
    __DEV__ && console.info(`Facebook Auth Failed - ${behavior}: `, caughtError.message || caughtError);

    return null;
  }

  return auth;
};
const fbLogin = async() => {
  let auth;
  const method = 'facebook';
  const cleanForce = (typeof process.env.FB_LOGIN_FORCE === 'string' && process.env.FB_LOGIN_FORCE.toLowerCase()) || '';

  if (cleanForce === 'both' || !cleanForce) {
    auth = await fbLoginWebview('native');
    if (!auth) {
      auth = await fbLoginWebview('web');
    }
  } else if (['native', 'web'].includes(cleanForce)) {
    auth = await fbLoginWebview(cleanForce);
  } else {
    // let OS/Device decide -> basically if(cleanForce === 'none'){}
    auth = await fbLoginWebview();
  }
  
  return auth;
}
export default fbLogin;
```

**Comment by tiffanym19** [comment]
I'm having also an issue like this, what is the update for this. Even the Facebook app installed in the ios device it opens the safari login instead.

**Comment by sun20171001** [comment]
Same issue

**Comment by vohoangankhanh** [comment]
Same problem

**Comment by emech-en** [comment]
Same problem

**Comment by mifi** [comment]
Had the same problem, my app would use the webview on iOS no matter if I had the Facebook app installed. For me what helped is to go to Safari, then log out of facebook. Now the webview 
 redirects to the Facebook app to login
